### ISSUE TITLE
no auto-completion for Dictionary keys

### ISSUE THREAD

**Issue by Libbae** [godotengine/godot#80145]
### Godot version

4.1.1

### System information

windows 10

### Issue description

there is no auto-completion for `Dictionary` keys in godot 4 (unless the `Dictionary` is a `const`)

i think this is a bug because there _**is**_ auto-completion for `Dictionary` keys in godot 3 and theres auto-completion when its a `const Dictionary`

### Steps to reproduce

```
var dictionary = { "try_to_autocomplete_this": "it wont work in godot 4" }
#const dictionary = { "try_to_autocomplete_this": "it WILL work in godot 4" }
func _ready():
	dictionary.try_to_
```

### Minimal reproduction project

N/A

**Comment by tokengamedev** [comment]
Kind of related to #78602. It seems there is pull request for the fix

**Comment by HolonProduction** [comment]
I don't think this really is a bug. It's how the type resolution works. If the dictionary isn't constant it could change at any point so when using it it is not guaranteed to have `try_to_autocomplete_this`. Therefore you should not access it in this way and do some check instead whether the dict has the key `try_to_autocomplete_this`.

Not quite sure what Godot 3 did here. But when it comes to types Godot 4 is just more type safe so we can't expect 1:1 compatibility.



> Kind of related to #78602. It seems there is pull request for the fix

I don't think this is related. In one case we are talking about auto completion in an dictionary declaration. Here we talk about accessing a dict. These things work differently.

**Comment by gnat** [comment]
Can confirm autocomplete works for pre-defined Dictionary keys in Godot 3, pretty convenient.

...but this convenience is broken in Godot 4- as OP describes.

> I don't think this really is a bug. It's how the type resolution works.

Autocomplete has **never been a guarantee** a script will work at runtime- never has been. It's just a best effort to provide reasonable contextual conveniences.

> Not quite sure what Godot 3 did here.

I suspect it parsed the script as usual and added existing text keys into the autocomplete system, which was very convenient when working with Dictionaries.

**Comment by HolonProduction** [comment]
While I stand by my opinion about this in isolation. I think it would be fine to accept a fix, because Dictionaries are heavily abused due to the absence of structs at the moment.

**Comment by Furgl** [comment]
I often use Enums as keys for Dictionaries of resources that I customize in the inspector. That way I can keep a collection of the resources and still be able to reference them individually using an Enum as the key.

If this issue is fixed, I would love to remove the Enums and just use keys (with proper auto-completion).

<details>
  <summary>Example</summary>
  
Code:  
```gdscript
# Old way: create enums for keys and reference dictionary values with them
enum Team { ONE, TWO }
@export var enums_teams: Dictionary[Team, TeamData]

func enums_print_teams() -> void:
	print("Team 1 name: "+enums_teams[Team.ONE].team_name)

# New way: add keys to lua-style dictionary and reference values with them
@export var lua_teams: Dictionary[String, TeamData] = { 
	ONE = null,
	TWO = null,
}

func lua_print_teams() -> void:
	print("Team 1 name: "+lua_teams.ONE.team_name)
  ```
Inspector:
<img width="486" height="409" alt="Image" src="https://github.com/user-attachments/assets/0202ac6d-6f5f-44ba-a018-b49228fd2268" />
</details>
